src/main.cpp

```diff
@@ -1598,17 +1602,29 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
             return error("AcceptToMemoryPool: BUG! PLEASE REPORT THIS! ConnectInputs failed against MANDATORY but not STANDARD flags %s", hash.ToString());
         }
 
-        // Store transaction in memory
-        pool.addUnchecked(hash, entry, !IsInitialBlockDownload(Params()));
+        {
+            // We lock to prevent other threads from accessing the mempool between adding and evicting
+            LOCK(pool.cs);
+            
+            // Store transaction in memory
+            pool.addUnchecked(hash, entry, !IsInitialBlockDownload(Params()));
+
+            // Add memory address index
+            if (fAddressIndex) {
+                pool.addAddressIndex(entry, view);
+            }
 
-        // Add memory address index
-        if (fAddressIndex) {
-            pool.addAddressIndex(entry, view);
-        }
+            // insightexplorer: Add memory spent index
+            if (fSpentIndex) {
+                pool.addSpentIndex(entry, view);
+            }
 
-        // insightexplorer: Add memory spent index
-        if (fSpentIndex) {
-            pool.addSpentIndex(entry, view);
+            // In normal circumstances the following should be boost::none
+            boost::optional<uint256> maybeDropTxId;
+            list<CTransaction> removed;
+            while ((maybeDropTxId = pool.ensureSizeLimit()).is_initialized()) {
+                pool.remove(pool.mapTx.find(maybeDropTxId.get())->GetTx(), removed, true);
```

Comment:
Why is the looping and removal not just part of `CTxMemPool::ensureSizeLimit`? As it is, the call does _not_ ensure the size limit is met, because the caller can choose to ignore the return value.